File: 15/CCC '15 S3 - Gates.cpp
root = [0] * 100001

def find(x): 
    if x!=root[x]:
        root[x] = find(root[x])
    return root[x]

def merge(x,y): 
    root[find(x)] = find(y)

g = int(input())
p = int(input())
for i in range(1,g+1):
    root[i] = i
i = 0
kontinue = True
while i < p and kontinue:
    wanted = int(input())
    root = find(wanted)
    if root == 0:
        print(i)
        kontinue = False 
    merge(root,root-1)
    i = i + 1
if kontinue:    
    print(p)

File: 15/CCC '15 S3 - Gates.py
# store free gates in a list
# assign each plane to the highest possible gate
# use binary search to find the highest possible gate
# if it is index 0, no gates left
# otherwise, remove that gate from the list 

import bisect

g = int(input())

p = int(input())

free = list(range(1, g+1))

ans = 0
for x in range(p):
    # print(free)
    plane = int(input())

    idx = bisect.bisect_right(free, plane)
    # print(idx)

    if idx == 0:
        break

    free.pop(idx-1)

    ans += 1

print(ans)
